### PR TITLE
Add "Optimizing" chapter to dynamic-dags section

### PR DESCRIPTION
**Pull request by potiuk**
There is a nice way to optimize dynamic DAG generation by our
users. Adding a short chapter linking to example on how this can
be done might guide them to do similar approach.

We handle several cases here:

* starting task via Python interpreter
* starting task via forking
* running "airflow tasks test" command

The detection here is rather complex and in the follow up PR
we will add a more robust detection mechanims (but it will be
only available as of Airflow 2.4)

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Review by feluelle — APPROVED** [review]
This is genius! 🚀

**Comment by potiuk** [comment]
I will wait for others to improve the English wording probably :)

**Comment by pierrejeambrun** [comment]
Very interesting, I think this will help a lot of people 💪

File: docs/apache-airflow/howto/dynamic-dag-generation.rst [hunk 1]
```diff
@@ -140,3 +140,20 @@ Each of them can run separately with related configuration
...
+
+
+Optimizing DAG parsing in workers/Kubernetes Pods
+-------------------------------------------------
+
+Sometimes when you generate a lot of Dynamic DAGs in single DAG file, it might cause unnecessary delays
+when the DAG file is parsed in worker or in Kubernetes POD. In Workers or Kubernetes PODs, you actually
+need only the single DAG (and even single Task of the DAG) to be instantiated in order to execute the task.
+If creating your DAG objects takes a lot of time, and each generated DAG is created independently from each
+other, this might be optimized away by simply skipping the generation of DAGs in worker.
```
Comment (BasPH) [inline docs/apache-airflow/howto/dynamic-dag-generation.rst]: Couple of suggestions for clarity/conciseness. Would also add a self-contained example, so that the reader can gather all information from just the docs.

```suggestion
Sometimes when you generate a lot of Dynamic DAGs in single DAG file, it might cause unnecessary delays
when the DAG file is parsed in worker or in Kubernetes POD. In Workers or Kubernetes PODs, you actually
need only the single DAG (and even single Task of the DAG) to be instantiated in order to execute the task.
If creating your DAG objects takes a lot of time, and each generated DAG is created independently from each
other, this might be optimized away by simply skipping the generation of DAGs in worker.
```

The parsing time of dynamically generated DAGs in Airflow workers can be optimized. This optimization is most effective when the number of generated DAGs is high. The Airflow scheduler requires loading of a complete DAG file to process all metadata. However, an Airflow worker requires only a single DAG object to execute a task. This allows us to skip the generation of unnecessary DAG objects in the worker, shortening the parsing time. Upon evaluation of a DAG file, command line arguments are supplied which we can use to determine whether the scheduler or worker evaluates the file:

- Scheduler args: ``["scheduler"]``
- Worker args: ``["airflow", "tasks", "run", "dag_id", "task_id", ...]``

Upon iterating over the collection of things to generate DAGs for, use these arguments to determine whether you need to generate all DAG objects (when running in the scheduler), or to generate only a single DAG object (when running in a worker):

.. code-block:: python
    :emphasize-lines: 1,2,3,7,8

    current_dag = None
    if len(sys.argv) > 3:
        current_dag = sys.argv[3]

    for thing in list_of_things:
        dag_id = f"generated_dag_{thing}"
        if current_dag is not None and current_dag != dag_id:
            continue  # skip generation of non-selected DAG
        
        dag = DAG(dag_id=dag_id, ...)
        globals()[dag_id] = dag

A nice example is shown in the
[Airflow's Magic Loop](https://medium.com/apache-airflow/airflows-magic-loop-ec424b05b629) blog post that describes how parsing in workers was reduced from 120 seconds to 200 ms.

File: docs/apache-airflow/howto/dynamic-dag-generation.rst [hunk 2]
```diff
@@ -140,3 +140,20 @@ Each of them can run separately with related configuration
 
 .. warning::
   Using this practice, pay attention to "late binding" behaviour in Python loops. See `that GitHub discussion <https://github.com/apache/airflow/discussions/21278#discussioncomment-2103559>`_ for more details
+
+
+Optimizing DAG parsing in workers/Kubernetes Pods
```
Comment (BasPH) [inline docs/apache-airflow/howto/dynamic-dag-generation.rst]: I don't think mentioning "Kubernetes Pods" here is correct? A pod is just the thing that runs a scheduler or worker?

File: docs/apache-airflow/howto/dynamic-dag-generation.rst (hunk 1 above)
Comment (potiuk) [reply docs/apache-airflow/howto/dynamic-dag-generation.rst]: Yep. Self contained example is something I also wanted to add after creating the PR :)

File: docs/apache-airflow/howto/dynamic-dag-generation.rst (hunk 2 above)
Comment (potiuk) [reply docs/apache-airflow/howto/dynamic-dag-generation.rst]: What I mean by that is really the K8S Pod run by the K8S Executor. "Worker" is more of the Celery term - it does not really apply to the KubernetesExecutor. But maybe "Airflow task Pod" would be better? WDYT?

Comment (BasPH) [reply docs/apache-airflow/howto/dynamic-dag-generation.rst]: This trick works regardless of the chosen executor, right?

Comment (eladkal) [reply docs/apache-airflow/howto/dynamic-dag-generation.rst]: Should we mention k8s here?
This is a general howto doc it also appears at the beginning of our documentation we can't expect users to be familiar with K8s here and the article contains several k8s references.
Maybe the k8s related should be extracted to the k8s executor section with a reference note?

Comment (potiuk) [reply docs/apache-airflow/howto/dynamic-dag-generation.rst]: > This trick works regardless of the chosen executor, right?

Correct.

> Should we mention k8s here?

Yeah. We should not. I think it's just mostly "namiing" problem. There is nothing K8S-specific here. We can simply mention "when your task is being executed"   rather than mentioning K8S POd or Worker. The thing is that people don't generaly realise that no-matter which executor they use, the task is actually executed with the right "airflow run task" command - for most people this is an implementation detail that they do not know. But in this case we are talking about optimisation which is on such a level, that educating the users about it, mighy make sense. Let me rephrase it.

Comment (BasPH) [reply docs/apache-airflow/howto/dynamic-dag-generation.rst]: Hmm I tested this on Local, Celery, and Kubernetes-executor and it only seems to work on the KubernetesExecutor. It seems args are passed differently on Local and Celery-executor. This is the DAG I'm using:

```python
import datetime
import logging
import sys

from airflow import DAG
from airflow.operators.bash import BashOperator

logger = logging.getLogger(__name__)

current_dag = None
if len(sys.argv) > 3:
    current_dag = sys.argv[3]

for thing in range(100):
    dag_id = f"generated_dag_{thing}"
    logger.info(f"CURRENT DAG = {current_dag}, DAG_ID = {dag_id}, sys.argv = {sys.argv}")
    if current_dag is not None and current_dag != dag_id:
        continue

    print(f"GENERATING DAG {dag_id}. Args = {sys.argv}")
    with DAG(dag_id=dag_id, schedule_interval="@daily", start_date=datetime.datetime(2022, 7, 1)) as dag:
        BashOperator(task_id="hello", bash_command="echo bla")
        globals()[dag_id] = dag
```

Here's a snippet of the output on CeleryExecutor:

```text
[2022-07-18 13:46:24,114: WARNING/ForkPoolWorker-15] GENERATING DAG generated_dag_65. Args = ['/home/airflow/.local/bin/airflow', 'celery', 'worker']
[2022-07-18 13:46:24,116: INFO/ForkPoolWorker-15] CURRENT DAG = None, DAG_ID = generated_dag_66, sys.argv = ['/home/airflow/.local/bin/airflow', 'celery', 'worker']
```

And on KubernetesExecutor:

```text
[2022-07-18 13:39:56,709] {scheduler_worker_distinction.py:20} INFO - CURRENT DAG = generated_dag_1, DAG_ID = generated_dag_99, sys.argv = ['/usr/local/bin/airflow', 'tasks', 'run', 'generated_dag_1', 'hello', 'scheduled__2022-07-13T00:00:00+00:00', '--local', '--subdir', 'DAGS_FOLDER/scheduler_worker_distinction.py']
```

I like this optimization, but if it only works on the KubernetesExecutor we should be explicit about it.

Comment (potiuk) [reply docs/apache-airflow/howto/dynamic-dag-generation.rst]: Hmm. I was quite sure celery worker is also doing "airflow run task" under the hood - but yeah. I think with pre-fork model on Celery it behaves like that. 

I think it still can be an optimised also for celery but yeah. We should work out the "generic" way of distinguishing between "scheduler" / "dag processor" and "task execution"

Comment (potiuk) [reply docs/apache-airflow/howto/dynamic-dag-generation.rst]: Of course `execute_tasks_new_python_interpreter  = False` -> is what causes this not to work :) - we just need to handle this case.

Comment (potiuk) [reply docs/apache-airflow/howto/dynamic-dag-generation.rst]: I think we could either get the command from "getproctitle" (that would likely work even now) or maybe we should introduce a more robust way (setting env variables AIRFLOW_CURRENT_DAG_ID AIRFLOW_CURRENT_TASK_ID?) to make it really robust in the future ?

**Comment by potiuk** [comment]
ocnverted to draft to iterate a bit more on it

File: docs/apache-airflow/howto/dynamic-dag-generation.rst [hunk 3]
```diff
@@ -140,3 +140,74 @@ Each of them can run separately with related configuration
...
+Sometimes when you generate a lot of Dynamic DAGs from a single DAG file, it might cause unnecessary delays
+when the DAG file is parsed during task execution. This is visible as delay before a task starts
+and when it gets executed. Sometimes the delays might be minutes.
+
+Why is this happening ?
+
+You might not be aware but when your task is executed just before execution, airflow parses the Python
+file the dag comes from. When tasks are executed, parsing time of dynamically generated DAGs
+when task are executed can be optimized. This optimization is most effective when the
+number of generated DAGs is high.
```
Comment (jedcunningham) [inline docs/apache-airflow/howto/dynamic-dag-generation.rst]:
```suggestion
Why is this happening? You might not be aware but just before your task is executed, airflow parses
the Python file the dag comes from.
```

I think these last 2 sentences are here too early in this section.

File: docs/apache-airflow/howto/dynamic-dag-generation.rst [hunk 4]
```diff
@@ -140,3 +140,74 @@ Each of them can run separately with related configuration
...
+
+The Airflow Scheduler (or DAG Processor) requires loading of a complete DAG file to process all metadata.
+However, task execution requires only a single DAG object to execute a task. Knowing this, allows us to
```
Comment (jedcunningham) [inline docs/apache-airflow/howto/dynamic-dag-generation.rst]:
```suggestion
However, task execution requires only a single DAG object to execute a task. Knowing this, we can
```

File: docs/apache-airflow/howto/dynamic-dag-generation.rst [hunk 5]
```diff
@@ -140,3 +140,74 @@ Each of them can run separately with related configuration
...
+skip the generation of unnecessary DAG objects when task is executed, shortening the parsing time.
+Upon evaluation of a DAG file, command line arguments are supplied which we can use to determine which
```
Comment (jedcunningham) [inline docs/apache-airflow/howto/dynamic-dag-generation.rst:166]:
```suggestion
This optimization is most effective when the number of generated DAGs is high.

Upon evaluation of a DAG file, command line arguments are supplied which we can use to determine which
```

This is a better place for it I think.

File: docs/apache-airflow/howto/dynamic-dag-generation.rst [hunk 6]
```diff
@@ -140,3 +140,74 @@ Each of them can run separately with related configuration
...
+  for thing in list_of_things:
+      dag_id = f"generated_dag_{thing}"
+      if current_dag is not None and current_dag != dag_id:
+          continue  # skip generation of non-selected DAG
+
+      dag = DAG(dag_id=dag_id, ...)
+      globals()[dag_id] = dag
+
+
+Checking if the second argument is ``tasks`` also allows to optimize ``airflow tasks test`` command
```
Comment (jedcunningham) [inline docs/apache-airflow/howto/dynamic-dag-generation.rst]:
```suggestion
Checking if the second argument is ``tasks`` also allows us to optimize the ``airflow tasks test`` command
```

**Review by jedcunningham — COMMENTED** [review]
I'm a little torn on documenting this in the official docs. Almost feels like if we are going to officially suggest it, we should have a better DAG authoring experience for it.

File: docs/apache-airflow/howto/dynamic-dag-generation.rst [hunk 7]
```diff
@@ -140,3 +140,74 @@ Each of them can run separately with related configuration
...
+
+
+Optimizing DAG parsing delays during execution
+----------------------------------------------
+
+Sometimes when you generate a lot of Dynamic DAGs from a single DAG file, it might cause unnecessary delays
+when the DAG file is parsed during task execution. This is visible as delay before a task starts
+and when it gets executed. Sometimes the delays might be minutes.
+
+Why is this happening ?
```
Comment (BasPH) [inline docs/apache-airflow/howto/dynamic-dag-generation.rst]:
```suggestion
Why is this happening?
```

File: docs/apache-airflow/howto/dynamic-dag-generation.rst [hunk 8]
```diff
@@ -140,3 +140,74 @@ Each of them can run separately with related configuration
...
 .. warning::
   Using this practice, pay attention to "late binding" behaviour in Python loops. See `that GitHub discussion <https://github.com/apache/airflow/discussions/21278#discussioncomment-2103559>`_ for more details
+
+
+Optimizing DAG parsing delays during execution
+----------------------------------------------
+
+Sometimes when you generate a lot of Dynamic DAGs from a single DAG file, it might cause unnecessary delays
+when the DAG file is parsed during task execution. This is visible as delay before a task starts
+and when it gets executed. Sometimes the delays might be minutes.
```
Comment (BasPH) [inline docs/apache-airflow/howto/dynamic-dag-generation.rst]:
```suggestion
when the DAG file is parsed during task execution. The impact is a delay before a task starts.
```

Would change the wording from "visible" to "impact" since there isn't any process actively notifying a user of task delay. Also, would remove the sentence "Sometimes the delays might be minutes." since it depends on a user's DAG code and the delay could be anything.

File: docs/apache-airflow/howto/dynamic-dag-generation.rst [hunk 9]
```diff
@@ -140,3 +140,74 @@ Each of them can run separately with related configuration
...
+
+Why is this happening ?
+
+You might not be aware but when your task is executed just before execution, airflow parses the Python
```
Comment (BasPH) [inline docs/apache-airflow/howto/dynamic-dag-generation.rst]:
```suggestion
You might not be aware but when your task is executed just before execution, Airflow parses the Python
```

File: docs/apache-airflow/howto/dynamic-dag-generation.rst [hunk 10]
```diff
@@ -140,3 +140,74 @@ Each of them can run separately with related configuration
...
+file the dag comes from. When tasks are executed, parsing time of dynamically generated DAGs
```
Comment (BasPH) [inline docs/apache-airflow/howto/dynamic-dag-generation.rst]:
```suggestion
file the DAG comes from. When tasks are executed, parsing time of dynamically generated DAGs
```

File: docs/apache-airflow/howto/dynamic-dag-generation.rst [hunk 11]
```diff
@@ -140,3 +140,74 @@ Each of them can run separately with related configuration
...
+when task are executed can be optimized. This optimization is most effective when the
```
Comment (BasPH) [inline docs/apache-airflow/howto/dynamic-dag-generation.rst]:
```suggestion
when a task is executed can be optimized. This optimization is most effective when the
```

File: docs/apache-airflow/howto/dynamic-dag-generation.rst (hunk 4 above)
Comment (BasPH) [inline docs/apache-airflow/howto/dynamic-dag-generation.rst]:
```suggestion
However, task execution requires only a single DAG object to execute a task. Knowing this allows us to
```

File: docs/apache-airflow/howto/dynamic-dag-generation.rst [hunk 12]
```diff
@@ -140,3 +140,74 @@ Each of them can run separately with related configuration
...
+number of generated DAGs is high.
+
+The Airflow Scheduler (or DAG Processor) requires loading of a complete DAG file to process all metadata.
+However, task execution requires only a single DAG object to execute a task. Knowing this, allows us to
+skip the generation of unnecessary DAG objects when task is executed, shortening the parsing time.
+Upon evaluation of a DAG file, command line arguments are supplied which we can use to determine which
+airflow component performs parsing:
```
Comment (BasPH) [inline docs/apache-airflow/howto/dynamic-dag-generation.rst]:
```suggestion
Airflow component performs parsing:
```

File: docs/apache-airflow/howto/dynamic-dag-generation.rst [hunk 13]
```diff
@@ -140,3 +140,74 @@ Each of them can run separately with related configuration
...
+
+* Scheduler/DAG Processor args: ``["scheduler"]`` or ``["dag-processor"]``
```
Comment (BasPH) [inline docs/apache-airflow/howto/dynamic-dag-generation.rst]: Would start args with `airflow` since that's also shown in task execution args.

File: docs/apache-airflow/howto/dynamic-dag-generation.rst [hunk 14]
```diff
@@ -140,3 +140,74 @@ Each of them can run separately with related configuration
...
+* Task execution args: ``["airflow", "tasks", "run", "dag_id", "task_id", ...]``
+
+However, depending on the executor used and forking model, those args might be available via ``sys.args``
+or via name of the process running. Airflow either executes tasks via running a new Python interpreter or
+sets the name of the process as "airflow task supervisor: {ARGS}".
+
+Upon iterating over the collection of things to generate DAGs for, you can use these arguments to determine
+whether you need to generate all DAG objects (when parsing in the DAG File processor), or to generate only
+a single DAG object (when executing the task).
```
Comment (BasPH) [inline docs/apache-airflow/howto/dynamic-dag-generation.rst]:
```suggestion
a single DAG object (when executing the task):
```

File: docs/apache-airflow/howto/dynamic-dag-generation.rst [hunk 15]
```diff
@@ -140,3 +140,74 @@ Each of them can run separately with related configuration
...
+
+
+.. code-block:: python
+  :emphasize-lines: 1,2,3,7,8
```
Comment (BasPH) [inline docs/apache-airflow/howto/dynamic-dag-generation.rst]: Change to highlight the relevant lines

File: docs/apache-airflow/howto/dynamic-dag-generation.rst [hunk 16]
```diff
@@ -140,3 +140,74 @@ Each of them can run separately with related configuration
...
+      try:
+          PROCTITLE_PREFIX = "airflow task supervisor: "
+          proctitle = str(setproctitle.getproctitle())
+          if proctitle.startswith(PROCTITLE_PREFIX):
+              args_string = proctitle[len(PROCTITLE_PREFIX) :]
+              args = ast.literal_eval(args_string)
+              if len(args) > 3 and args[1] == "tasks":
+                  current_dag = args[3]
+      except:
+          pass
```
Comment (BasPH) [inline docs/apache-airflow/howto/dynamic-dag-generation.rst]: IMO this if/else could use a comment e.g. "args supplied to process" and "process is forked and task args are supplied via proctitle".

File: docs/apache-airflow/howto/dynamic-dag-generation.rst [hunk 17]
```diff
@@ -140,3 +140,74 @@ Each of them can run separately with related configuration
...
+
+  for thing in list_of_things:
+      dag_id = f"generated_dag_{thing}"
+      if current_dag is not None and current_dag != dag_id:
+          continue  # skip generation of non-selected DAG
+
+      dag = DAG(dag_id=dag_id, ...)
+      globals()[dag_id] = dag
+
```
Comment (BasPH) [inline docs/apache-airflow/howto/dynamic-dag-generation.rst:219]: Let's show all imports in this code to make it an easily copy-pastable example.

File: docs/apache-airflow/howto/dynamic-dag-generation.rst [hunk 18]
```diff
@@ -140,3 +140,74 @@ Each of them can run separately with related configuration
...
+
+Checking if the second argument is ``tasks`` also allows to optimize ``airflow tasks test`` command
+in similar way, while it will also rule out accidental detection in case more arguments
+are passed to ``scheduler`` or ``dag-processor`` command.
```
Comment (BasPH) [inline docs/apache-airflow/howto/dynamic-dag-generation.rst]: IMO this sentence is unneccesary to a reader of the docs.

**Review by BasPH — COMMENTED** [review]
Overall LGTM, nice that the example works for both new and forked processes. Made some smaller comments.

File: docs/apache-airflow/howto/dynamic-dag-generation.rst (hunk 13 above)
Comment (potiuk) [reply docs/apache-airflow/howto/dynamic-dag-generation.rst]: Right.

File: docs/apache-airflow/howto/dynamic-dag-generation.rst (hunk 16 above)
Comment (potiuk) [reply docs/apache-airflow/howto/dynamic-dag-generation.rst]: Yeah. Comments are needed but rather than repeating in the comment what we actually do, it should explain why we are doing it:

```
# task executed by starting a new Python interpreter
```
and

```
 # task executed via forked process
````

are better comments in this case :)

File: docs/apache-airflow/howto/dynamic-dag-generation.rst (hunk 17 above)
Comment (potiuk) [reply docs/apache-airflow/howto/dynamic-dag-generation.rst:219]: Yeah. missed setproctitle and DAG.

File: docs/apache-airflow/howto/dynamic-dag-generation.rst (hunk 18 above)
Comment (potiuk) [reply docs/apache-airflow/howto/dynamic-dag-generation.rst]: Agree it feels more appropriate for commit message :)

**Comment by potiuk** [comment]
I addressed all comments I think -> Would love one more pass @BasPH @jedcunningham

File: docs/apache-airflow/howto/dynamic-dag-generation.rst [hunk 19]
```diff
@@ -140,3 +140,74 @@ Each of them can run separately with related configuration
...
+
+However, depending on the executor used and forking model, those args might be available via ``sys.args``
+or via name of the process running. Airflow either executes tasks via running a new Python interpreter or
+sets the name of the process as "airflow task supervisor: {ARGS}".
+
+This optimization is most effective when the number of generated DAGs is high.
+
+Upon iterating over the collection of things to generate DAGs for, you can use these arguments to determine
+whether you need to generate all DAG objects (when parsing in the DAG File processor), or to generate only
+a single DAG object (when executing the task)"
```
Comment (BasPH) [inline docs/apache-airflow/howto/dynamic-dag-generation.rst]:
```suggestion
a single DAG object (when executing the task):
```

File: docs/apache-airflow/howto/dynamic-dag-generation.rst [hunk 20]
```diff
@@ -140,3 +140,74 @@ Each of them can run separately with related configuration
...
+* Task execution args: ``["airflow", "tasks", "run", "dag_id", "task_id", ...]``
+
+
+When tasks are executed, parsing time of dynamically generated DAGs when a task is executed can be optimized.
+
+However, depending on the executor used and forking model, those args might be available via ``sys.args``
+or via name of the process running. Airflow either executes tasks via running a new Python interpreter or
+sets the name of the process as "airflow task supervisor: {ARGS}".
+
+This optimization is most effective when the number of generated DAGs is high.
```
Comment (BasPH) [inline docs/apache-airflow/howto/dynamic-dag-generation.rst]:
```suggestion
When tasks are executed, parsing time of dynamically generated DAGs can be optimized. This optimization is most effective when the number of generated DAGs is high.

However, depending on the executor used and forking model, those args might be available via ``sys.args``
or via name of the process running. Airflow either executes tasks via running a new Python interpreter or
sets the name of the process as "airflow task supervisor: {ARGS}".
```

Remove duplicate "When tasks are executed" + merge with the sentence below for better structure.

**Review by BasPH — COMMENTED** [review]
Few minor comments, otherwise LGTM!

File: docs/apache-airflow/howto/dynamic-dag-generation.rst [hunk 21]
```diff
@@ -140,3 +140,74 @@ Each of them can run separately with related configuration
...
+      try:
+          PROCTITLE_PREFIX = "airflow task supervisor: "
+          proctitle = str(setproctitle.getproctitle())
+          if proctitle.startswith(PROCTITLE_PREFIX):
+              # task executed via forked process
+              args_string = proctitle[len(PROCTITLE_PREFIX) :]
+              args = ast.literal_eval(args_string)
+              if len(args) > 3 and args[1] == "tasks":
+                  current_dag = args[3]
+      except:
```
Comment (csp33) [inline docs/apache-airflow/howto/dynamic-dag-generation.rst]:
```suggestion
      except Exception:
```

File: docs/apache-airflow/howto/dynamic-dag-generation.rst (hunk 20 above)
Comment (potiuk) [reply docs/apache-airflow/howto/dynamic-dag-generation.rst]: I shortened it a bit more - the optimization part was already mentioned above.

File: docs/apache-airflow/howto/dynamic-dag-generation.rst [hunk 22]
```diff
@@ -140,3 +140,78 @@ Each of them can run separately with related configuration
...
   Using this practice, pay attention to "late binding" behaviour in Python loops. See `that GitHub discussion <https://github.com/apache/airflow/discussions/21278#discussioncomment-2103559>`_ for more details
+
+
+Optimizing DAG parsing delays during execution
+----------------------------------------------
+
+Sometimes when you generate a lot of Dynamic DAGs from a single DAG file, it might cause unnecessary delays
+when the DAG file is parsed during task execution. The impact is a delay before a task starts.
+
+Why is this happening? You might not be aware but when your task is executed just before execution,
```
Comment (jedcunningham) [inline docs/apache-airflow/howto/dynamic-dag-generation.rst]:
```suggestion
Why is this happening? You might not be aware but just before your task is executed,
```

File: docs/apache-airflow/howto/dynamic-dag-generation.rst [hunk 23]
```diff
@@ -140,3 +140,78 @@ Each of them can run separately with related configuration
...
+Airflow parses the Python file the DAG comes from.
+
+The Airflow Scheduler (or DAG Processor) requires loading of a complete DAG file to process all metadata.
+However, task execution requires only a single DAG object to execute a task. Knowing this, we can
+skip the generation of unnecessary DAG objects when task is executed, shortening the parsing time.
```
Comment (jedcunningham) [inline docs/apache-airflow/howto/dynamic-dag-generation.rst]:
```suggestion
skip the generation of unnecessary DAG objects when a task is executed, shortening the parsing time.
```

**Comment by potiuk** [comment]
I know @jedcunningham you had reservation about putting that out now.  I thought a bit about it and I added one more chapter about "experimental" status of the practice - with lots of reservations and disclaimers.

If you think this is fine now @jedcunningham  @BasPH - I will  look at the robust solution in #25161  (and then we can think if we want to add any kind of __future__ approach. 

I think there is very little harm in adding the documentation about it even now with the "complex" code - as this will help some users and they will at least know that there is a  possibility of such optimization. 

The problem is real and serious - the more dynamic dags people will use, the more similiar solution  will be needed, so I prefer even to get some issues from people who tried and fail, so that we can possibly see if we can optimize it even further or maybe even add some other solutions in the future.

**Comment by jedcunningham** [comment]
The new paragraph looks good, and it does give us something to point at to say "we told you it was experimental", I'm not sure it changes much.

I'm not exactly sure how to articulate my thoughts, but I'll try.

This feels to me like we are helping hide the fact people didn't architect their DAGs properly. If you've hit the point where this is necessary, then you've already strayed out of the norm. Putting out a complex chuck of code that'll be copy pasted around, getting out of date, what have you is just adding an extra burden to maintenance and support. I can totally imagine down the road an upgrade breaking an early version of this and us getting hard to diagnose bug reports because of it. Worse, if the users DAGs are big/complex enough to need this, they likely can't/won't share it as a reproduction so that adds an extra burden.

There are other ways to DRY up a DAG than to have them all come from a single loop 🤷‍♂️.

I have no problem with folks opting in to this, knowing (hopefully) exactly the risks that come with it. However once we toss it in the official docs (even with a disclaimer paragraph), it becomes "officially blessed" to an extent, and I'm not sure we should do that for this iteration.

tldr: very clever solution, but too esoteric/brittle to be officially documented for end users imo.

**Comment by potiuk** [comment]
> The new paragraph looks good, and it does give us something to point at to say "we told you it was experimental", I'm not sure it changes much.
> 
> I'm not exactly sure how to articulate my thoughts, but I'll try.

I think you managed very well :).


> 
> This feels to me like we are helping hide the fact people didn't architect their DAGs properly. If you've hit the point where this is necessary, then you've already strayed out of the norm. Putting out a complex chuck of code that'll be copy pasted around, getting out of date, what have you is just adding an extra burden to maintenance and support. I can totally imagine down the road an upgrade breaking an early version of this and us getting hard to diagnose bug reports because of it. Worse, if the users DAGs are big/complex enough to need this, they likely can't/won't share it as a reproduction so that adds an extra burden.
> 

Yeah I know exactly what you mean. I have reservations myself, looking brittleness of the solution. I actually think I will do it differently. I will - myself - write a short blog post acompanying the one from @itayB where I add a more complete solution  and I also publish it in Airflow Publication.

> 
> There are other ways to DRY up a DAG than to have them all come from a single loop man_shrugging.

Should we maybe then mention THIS instead? Maybe that is a better approach that we warn the users not to do it?

> 
> tldr: very clever solution, but too esoteric/brittle to be officially documented for end users imo.

Question: woudl you also object if just get the "official" approach from #25161 ? Or is it mostly the brittleness ?

**Comment by jedcunningham** [comment]
> > There are other ways to DRY up a DAG than to have them all come from a single loop man_shrugging.
> 
> Should we maybe then mention THIS instead? Maybe that is a better approach that we warn the users not to do it?

I like this idea. We should do it, even if/when we do add a better short circuit for the loop approach.

> Question: woudl you also object if just get the "official" approach from #25161 ? Or is it mostly the brittleness ?

`get_parsing_context` seems much better and I wouldn't object to documenting that. It's tested and can be a stable interface for DAG authors. (Though I haven't looked closely at that PR and likely won't be able to until later next week)

> I will - myself - write a short blog post acompanying the one from @itayB where I add a more complete solution and I also publish it in Airflow Publication.

👍

**Comment by potiuk** [comment]
Closin this one then - blog post is being reviewed :)
